### PR TITLE
Add WindowsCaptureService module

### PR DESCRIPTION
**Pull request by pierricgimmig**
CaptureService is now platform-agnostic. Add LinuxCaptureService and
WindowsCaptureService modules which contain platform-specific code.

### CODE REVIEW

File: src/CaptureService/include/CaptureService/CaptureService.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+#include "CaptureStartStopListener.h"
+#include "OrbitBase/Result.h"
+#include "ProducerEventProcessor/GrpcClientCaptureEventCollector.h"
+#include "ProducerEventProcessor/ProducerEventProcessor.h"
+#include "services.grpc.pb.h"
+#include "services.pb.h"
+
+namespace orbit_capture_service {
+
+// CaptureService is an interface derived from the grpc capture service. It holds common
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/CaptureService/include/CaptureService/CaptureService.h:23]: Done.

File: src/CaptureService/CaptureService.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  bool new_insertion = capture_start_stop_listeners_.insert(listener).second;
+  CHECK(new_insertion);
+}
+
+void CaptureService::RemoveCaptureStartStopListener(CaptureStartStopListener* listener) {
+  bool was_removed = capture_start_stop_listeners_.erase(listener) > 0;
+  CHECK(was_removed);
+}
+
+ErrorMessageOr<void> CaptureService::InitializeCapture(
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/CaptureService/CaptureService.cpp:54]: Yep, done.

File: src/CaptureService/CommonProducerCaptureEventBuilders.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef CAPTURE_SERVICE_COMMON_PRODUCER_CAPTURE_EVENT_BUILDER_H_
+#define CAPTURE_SERVICE_COMMON_PRODUCER_CAPTURE_EVENT_BUILDER_H_
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/CaptureService/CommonProducerCaptureEventBuilders.cpp:6]: :-)

File: src/CaptureService/include/CaptureService/CaptureService.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+ protected:
+  void EstimateAndLogClockResolution();
+
+  ErrorMessageOr<void> InitializeCapture(
+      grpc::ServerReaderWriter<orbit_grpc_protos::CaptureResponse,
+                               orbit_grpc_protos::CaptureRequest>* reader_writer);
+  void TerminateCapture();
+
+  void WaitForStartCaptureRequestFromClient(
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/CaptureService/include/CaptureService/CaptureService.h:46]: Yep, cleaner.

File: src/CaptureService/include/CaptureService/CaptureService.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      grpc::ServerReaderWriter<orbit_grpc_protos::CaptureResponse,
+                               orbit_grpc_protos::CaptureRequest>* reader_writer,
+      orbit_grpc_protos::CaptureRequest& request);
+
+  void WaitForEndCaptureRequestFromClient(
+      grpc::ServerReaderWriter<orbit_grpc_protos::CaptureResponse,
+                               orbit_grpc_protos::CaptureRequest>* reader_writer,
+      orbit_grpc_protos::CaptureRequest& request);
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/CaptureService/include/CaptureService/CaptureService.h:54]: Ah perfect, done, thanks.

File: src/LinuxCaptureService/LinuxCaptureService.cpp [hunk 6]
```diff
@@ -110,104 +113,18 @@ static void StopInternalProducersAndCaptureStartStopListenersInParallel(
...
-  GrpcClientCaptureEventCollector client_capture_event_collector{reader_writer};
-  std::unique_ptr<ProducerEventProcessor> producer_event_processor =
-      ProducerEventProcessor::Create(&client_capture_event_collector);
-  TracingHandler tracing_handler{producer_event_processor.get()};
-  MemoryInfoHandler memory_info_handler{producer_event_processor.get()};
+  TracingHandler tracing_handler{producer_event_processor_.get()};
+  MemoryInfoHandler memory_info_handler{producer_event_processor_.get()};
 
   CaptureRequest request;
   reader_writer->Read(&request);
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/LinuxCaptureService/LinuxCaptureService.cpp:130]: ah yes of course, thanks, done.

File: src/CaptureService/include/CaptureService/CaptureService.h [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+  void StartEventProcessing(const orbit_grpc_protos::CaptureOptions& capture_options);
+  void FinalizeEventProcessing();
+
+ protected:
+  mutable absl::Mutex capture_mutex_;
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/CaptureService/include/CaptureService/CaptureService.h:60]: not needed here, removed.

File: src/CaptureService/include/CaptureService/CaptureService.h [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  virtual grpc::Status Capture(
+      grpc::ServerContext* context,
+      grpc::ServerReaderWriter<orbit_grpc_protos::CaptureResponse,
+                               orbit_grpc_protos::CaptureRequest>* reader_writer) = 0;
+
+  void AddCaptureStartStopListener(CaptureStartStopListener* listener);
+  void RemoveCaptureStartStopListener(CaptureStartStopListener* listener);
+
+ protected:
+  void EstimateAndLogClockResolution();
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/CaptureService/include/CaptureService/CaptureService.h]: done.

File: src/CaptureService/include/CaptureService/CaptureService.h [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  mutable absl::Mutex capture_mutex_;
+  bool is_capturing = false;
+
+  std::unique_ptr<orbit_producer_event_processor::GrpcClientCaptureEventCollector>
+      grpc_client_capture_event_collector_;
+  std::unique_ptr<orbit_producer_event_processor::ProducerEventProcessor> producer_event_processor_;
+
+  absl::flat_hash_set<CaptureStartStopListener*> capture_start_stop_listeners_;
+
+  uint64_t clock_resolution_ns_ = 0;
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/CaptureService/include/CaptureService/CaptureService.h:65]: done.

File: src/CaptureService/include/CaptureService/CaptureService.h [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                               orbit_grpc_protos::CaptureRequest>* reader_writer,
+      orbit_grpc_protos::CaptureRequest& request);
+
+  void StartEventProcessing(const orbit_grpc_protos::CaptureOptions& capture_options);
+  void FinalizeEventProcessing();
+
+ protected:
+  mutable absl::Mutex capture_mutex_;
+  bool is_capturing = false;
+
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/CaptureService/include/CaptureService/CaptureService.h]: done.

File: src/CaptureService/include/CaptureService/CaptureService.h [hunk 11]
```diff
@@ -14,39 +14,35 @@
...
 #include "OrbitBase/Logging.h"
 #include "OrbitBase/Profiling.h"
-#include "UserSpaceInstrumentation/InstrumentProcess.h"
 #include "services.grpc.pb.h"
 #include "services.pb.h"
 
 namespace orbit_capture_service {
 
-class CaptureServiceImpl final : public orbit_grpc_protos::CaptureService::Service {
+class CaptureService : public orbit_grpc_protos::CaptureService::Service {
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/CaptureService/include/CaptureService/CaptureService.h:25]: b/204595539

**Comment by pierricgimmig** [comment]
@florian-kuebler , @dpallotti , thanks once more for the reviews. I Addressed the remaining comments. I'll merge this now as it will unblock other PRs but I'd be happy to revisit if anything pops up.

File: src/CaptureService/include/CaptureService/CaptureService.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (dpallotti) [reply src/CaptureService/include/CaptureService/CaptureService.h:25]: As I wrote in the bug, I don't see a reason why the renaming (specifically of the class, but also of the module) can't already happen now. If then the module structure changes then we can rename the module later, but that shouldn't affect the current class name.

**Pull request merged**